src/coreComponents/managers/Functions/TableFunction.cpp

```diff
@@ -177,129 +179,47 @@ void TableFunction::reInitializeFunction()
       m_corners[jj][ii] = int(ii / pow( 2, jj )) % 2;
     }
   }
+
+  // Create the kernel wrapper
+  // Warning! This function must be called AFTER all the input parameters have been set (interpolation method, etc)
+  // TODO: see if there is a better way, this seems quite error prone
+  m_kernelWrapper =
+    std::make_unique< TableFunctionKernelWrapper >( m_interpolationMethod,
```

Comment:
I don't think there's a way to avoid having to recreate the kernel wrapper every time the table object is modified. Luckily, this doesn't happen much outside of unit tests (I think?). If only the coordinates and values change (but not the number of points in each dir), then array views will see the updated data automatically, but on any array resize the views need to be recreated.

You could get rid of the `unique_ptr` if you wanted to (if you gave a default constructor to `TableFunctionKernelWrapper`) and keep the wrapper object as a direct member of `TableFunction`. This would avoid an extra indirection in `Evaluate`. I'm ok with either solution though.

Reply:
Yes you are right I only had this problem in the `testFunctions` unit test. Now I recreate the kernel wrapper after each call to the setters.

After adding the default constructor to `TableFunctionKernelWrapper`, I did not find a good way to have the wrapper object as a direct member of `TableFunction` to remove the pointer:
- Currently, in the hpp file, I do a forward declaration of `class TableFunctionKernelWrapper`, and then there is the definition of the class `TableFunction`, and then that of the class `TableFunctionKernelWrapper`. So when I have the wrapper object as a direct member of `TableFunction`, the compiler says that `the field 'm_kernelWrapper' has incomplete type`.
- If I try the other approach: forward declare `class TableFunction`, then define the class `TableFunctionKernelWrapper`, and finally the class `TableFunction`, I get similar compilation problems with some types and variables needed in  `TableFunctionKernelWrapper`, like `TableFunction::InterpolationType` and `TableFunction::maxDimensions`.

Is there a simple workaround for these errors? (I left the pointer in the current version for now).

Reply:
Hmm, I see. Indeed, we have a circular dependency. The dependency of `TableFunction` on `TableFunctionKernelWrapper` is strong (does require the full definition if contains a member variable), but the other direction is weak (it only needs some nested constants and typedefs). Normally it would be resolved one of two ways:
 * taking those constants and typedefs out of `TableFunction`, which I already advised against (I definitely like the proper scoping of those);
 * or making `TableFunctionKernelWrapper` a nested (inner) class of `TableFunction`, which seems like a good idea to me, except we currently don't follow that pattern in any of the constitutive models that have a dedicated "update"/"kernel wrapper" class. In that case, it can lose the `TableFunction` part of its name and become simply `KernelWrapper`.

If you feel like trying the second option, go ahead. I can't think of a better solution yet. Leaving everything "as is" is fine too, I don't think fetching a `unique_ptr` is likely to have an observable effect, given the complexity of the function.

Reply:
I implemented the second option to deal with the circular dependency, with `KernelWrapper` now a nested class of `TableFunction`. I removed the `unique_ptr` and the `KernelWrapper` object is now a direct member of `TableFunction`. Thanks!